### ISSUE TITLE
rtorrent SCGI port or socket

### ISSUE THREAD

**Issue by zanloy** [linuxserver/docker-rutorrent#122]
I need the host system and possibly other docker containers to access the scgi port of rtorrent. This allows other containers (eg: radarr, sickchill) to add torrents. But even though the documentation says to expose port 5000, when you look at the rtorrent.rc file, it's using a socket and not port 5000. Unfortunately I can not enable the port in the config (it's commented out by default) because it fails saying that "SCGI already enabled." This seems to be a limitation of rtorrent. You can either use a socket or the port but not both.

So the solution to this is either
a) export the socket file in /config instead of /run/php/.rtorrent.sock
b) change to using tcp port and update rutorrent's config.php to point to tcp port instead of socket.

I want to know what the best way to move forward is. I do not want to make edits inside the container because that breaks my ability to bring down updates from upstream.

**Comment by zanloy** [comment]
I just realized that I can edit the rutorrent config.php outside the container so I can fix that. But you should probably just remove the references to exposing port 5000 if you don't use it anyways.

**Comment by flatline-studios** [comment]
Just wasted several hours discovering this the hard way... Thanks for posting this, or I might still of been wasting time.

**UPDATE**: Maybe just a note on the readme to mention that there's a limitation in rTorrent to only allow either the socket or the SCGI, and that the shipped/default rtorrent.rc file uses the socket option? That's why the `scgi_port` line is commented out.

**Comment by adamwinn** [comment]
@zanloy or @flatline-studios Can either of you share you `rtorrent.rc` and `config.php` config?  I can't seem to get mine to work.

**Comment by flatline-studios** [comment]
> @zanloy or @flatline-studios Can either of you share you `rtorrent.rc` and `config.php` config? I can't seem to get mine to work.

Hi, my issue was different to this one... What is it that you're trying to do exactly?

Quick bit of guidance... rTorrent itself uses either:
- A `socket` (which is a *file* that acts as a kind of hub that services can communicate through)
- An `scgi` service (which is a *protocol* [or language of sorts] that related services have agreed upon, and can both *speak*)

This docker image uses the `socket` approach in its default `rtorrent.rc`, and apparently rTorrent can only use one of these communication mechanisms at a time.

**Comment by zanloy** [comment]
So I went with a different route to get all the apps to talk to each other. You can't expose the socket so I setup rtorrent to use 5000/tcp. Then I created a docker network called 'backend' and added all my containers that needed to talk to each other on that network. Because by default docker-compose creates isolated networks for each compose file. This means all containers in a single docker-compose.yml can talk to each other using their hostnames. Another gotcha is that docker has a 'default' network that everything is added to if a custom network is not defined but that network *does not* allow containers to talk to each other. You have to create the custom network. Here's an example:

rutorrent/docker-compose.yml:
```
---
version: "2"

services:
  rutorrent:
    image: linuxserver/rutorrent:latest
    container_name: rutorrent
    environment:
      - PUID=1000
      - PGID=1000
    volumes:
      - ./config:/config
      - /home/zan/downloads:/downloads
    ports:
      - "6980:80"
      - "51413:51413"
      - "6881:6881/udp"
    networks:
      - backend
    restart: unless-stopped

networks:
  backend:
    external: true
```

sickchill/docker-compose.yml:
```
---
version: "3"

services:
  sickchill:
    image: linuxserver/sickchill:latest
    container_name: sickchill
    environment:
      - PUID=1000
      - PGID=1000
      - TZ=America/New_York
    volumes:
      - ./config:/config
      - /netshare/downloads:/downloads
      - /netshare/videos/tv:/tv
      - /netshare/videos/kids/tv:/kids-tv
    ports:
      - "7280:8081"
    networks:
      - backend
    restart: unless-stopped

networks:
  backend:
    external: true
```

This is not a current working config though for rtorrent because I actually moved all my stuff like sickchill/radarr/etc to use usenet instead of torrents and sabnzbd to download files. But when it was working, you could then just use http://rutorrent:5000/RPC2 as your url for the configurations and it'd talk to the correct container.

**Comment by zanloy** [comment]
btw, you will have to edit the rtorrent.rc file to use tcp *and* edit rutorrent's config.php to look at localhost:5000 instead of the socket file. This wasn't as easy as it sounded when I tried it. It was hours of trial and error. Unfortunately since I had to reinstall everything and no longer use torrents, I don't have the exact working config you guys could use.

**Comment by gmmoreira** [comment]
It's actually pretty easy to access the RPC endpoint. You don't need to modify any config files. Unfortunately, there's no mention of that in the README.

Rtorrent by default is using a SOCK file for SCGI instead of binding a TCP port. Rutorrent by default uses this sock file directly.

Then we have Nginx, acting as a reverse proxy for Rutorent. It also provides a location (/RPC2), doing scgi_pass to the rtorrent sock file.

This is the nginx conf at the time of writing:

```
upstream backendrtorrent {
    server unix:/run/php/.rtorrent.sock;
}

server {
    listen 80 default_server;

    listen 443 ssl;

    root /app/rutorrent;
    index index.php index.html index.htm;

    server_name _;

    ssl_certificate /config/keys/cert.crt;
    ssl_certificate_key /config/keys/cert.key;

    client_max_body_size 0;

    location / {
        location ~ .php$ {
            fastcgi_split_path_info ^(.+\.php)(.*)$;
            fastcgi_pass 127.0.0.1:9000;
            fastcgi_index index.php;
            include /etc/nginx/fastcgi_params;
        }
    }

    location /RPC2 {
        access_log /config/log/nginx/rutorrent.rpc2.access.log;
        error_log /config/log/nginx/rutorrent.rpc2.error.log;
        include /etc/nginx/scgi_params;
        scgi_pass backendrtorrent;
    }
}
```

Basically, to access the RPC endpoint, you don't need to use port 5000 or anything. Just use the same port binded in the docker run/compose and mount point /RPC2. You can also use port 443 for SSL support.
